### DISCUSSION TITLE
Using Storybook to test component interactions

### DISCUSSION THREAD

**Discussion by zarahzachz** [CMSgov/design-system#1800] [Ideas]
Found [a post from Storybook](https://storybook.js.org/blog/test-component-interactions-with-storybook/) showing what a marriage between it and Testing Library would look like. 

This part got my attention:
> Testing interactions is a widespread pattern for verifying user behavior. You provide mock data to set up a test scenario, simulate user interactions using [Testing Library](https://testing-library.com/), and check the resultant DOM structure.
> 
> In Storybook, **this familiar workflow happens in your browser**. That makes it easier to debug failures because you're running tests in the same environment as you develop components—the browser.

![image](https://user-images.githubusercontent.com/5159392/167206810-7f384a8f-66d8-4359-85a5-14515506a19a.png)

I'm wondering if this is something we've looked at or could consider doing discovery on?

**Comment by pwolfert** [comment]
This is intriguing.
1. I'm curious how performant it is. Running all our unit tests triggers flashbacks about Karma. I wouldn't want them to be flaky, because I have yet to come across a browser-testing suite that isn't at least a little flaky, no matter how hard you try to write flexible tests.
2. I do like that it uses similar syntax as `testing-library`, but does `@storybook/testing-library` actually use testing-library or just conform to its API? I like the idea of it being built off of proven technologies
3. I love that it will run with a CLI with Jest but that it'll give you a link to see the test running visually in a browser! ![failing test in Jest links to test in browser](https://storybookblog.ghost.io/content/images/2022/04/click-to-debug.gif)

**Comment by pwolfert** [comment]
Two more things:
1. I wonder if all these interaction-test stories would show up in our main Storybook (might be clutter)
2. We already build and run Storybook in our CI tests because of our Storybook-based visual regression tests, so at least we wouldn't be adding the time to build storybook to the execution time for our CI tests (which are already getting kind of long). We'd start Storybook for core, run our interaction tests, leave it running and run loki (VRTs), and then shut it down...oh, but we actually recently built in the storybook-running part into our loki command for convenience...and I wouldn't want to make that _less_ convenient 🤔
